### PR TITLE
Automatically detect whether the terminal background is a light color

### LINKED ISSUE #28690: Default REPL output is difficult to read on light background
I'm using Julia on macOS, and my terminals have a light (almost white) background. The default color scheme of the REPL is difficult to read there, as e.g. yellow (for warnings) is almost unreadable.

[This page](https://unix.stackexchange.com/questions/245378/common-environment-variable-to-set-dark-or-light-terminal-background) has some hints for how to automatically detect whether the terminal background is light or dark.

### PR DESCRIPTION
**Pull request by eschnett**
This detects (or tries to) whether the terminal background is a light color. The REPL then chooses a different default color scheme that is easier to read, avoiding e.g. yellow.

This only affects the default colors; any user-chosen colors remain the same.

Closes #28690.

### CODE REVIEW

**Comment by KristofferC** [comment]
What is the user has changed the way the default colours are shown in the terminal to already fix this issue? Then we will still swap to these non-standard colors. 

If the defaults in some terminals make some colors unreadable, maybe it is the defaults that need to be changed?

File: stdlib/REPL/src/Terminals.jl [hunk 1]
```diff
@@ -169,6 +171,31 @@ else
...
+        # for the 16 possible colors
+        colorbg = tryparse(Int, colorfgbg[last(sep)+1:end])
+        if colorbg !== Nothing && colorbg >= 0 && colorbg <= 15
+            return colorbg == 7 || colorbg >= 9
+        end
+    end
+    # (We could/should use more mechanisms here)
+    # Blindly guess an OS-dependent default
+    if Sys.isapple()
+        return true
```
Comment (Keno) [inline stdlib/REPL/src/Terminals.jl]: This is a bad assumption. iTerm2 defaults to black. However, iTerm2 does have color query escape codes.

Comment (eschnett) [reply stdlib/REPL/src/Terminals.jl]: To my knowledge, iTerm2 sets the `COLORFGBG` environment variable that we check above, thus this default isn't reached.

**Comment by eschnett** [comment]
@KristofferC If the user sets colors, then Julia won't use its default colors. This change only modifies the defaults that Julia uses if there is no customization.

I don't think it is possible to choose colors that work with every background. Light colors are unreadable on light backgrounds, and dark colors are unreadable on dark backgrounds. Plus, Julia wants to use several different colors to improve the output.

**Comment by StefanKarpinski** [comment]
This seems like a good direction to make the default experience better. But does this light/dark background detection really work? It seems kind of implausible to me that it would actually work.

**Comment by eschnett** [comment]
Appveyor reports an all-green status, but the Github status here on this page doesn't update.

File: stdlib/REPL/src/Terminals.jl (hunk 1 above)
Comment (KristofferC) [reply stdlib/REPL/src/Terminals.jl]: I don't think we should ever default this to true solely based on OS.

Comment (eschnett) [reply stdlib/REPL/src/Terminals.jl]: I removed the default based on OS; all OS now use the same default

**Comment by eschnett** [comment]
The patch now uses roughly the same mechanism as vim. Querying the terminal via ANSI escape sequences work for iTerm2 and xterm (and probably their derivatives). There are also command line options to override things (`--color=[auto | on | off | light | dark]`). Ready for review.

**Comment by StefanKarpinski** [comment]
Bump, who should review?

**Comment by eschnett** [comment]
I now find that many Julia packages set colours directly (e.g. by using `:yellow`), and are not going through either environment variables or a colour scheme (as the REPL does). It seems that it would be a huge task to change this, as it would involve changing many packages as well (Pkg, profiler, ..., ?), and maybe even setting up usability coding guidelines. (Or maybe Julia should set the terminal background colour as well, to ensure that everybody is using the same background colour and thus has readable colours.)

A work-around would be to conditionally change the definition of `:yellow` to be `:orange` instead, i.e. to change the meaning of colours behind package writers' back. That goes in the same direction as @KristofferC's comment, except that he changes the terminal settings.

In summary, I'm disappointed that my pull request didn't actually solve the problem it was supposed to solve. I don't know whether it's actually useful in practice. If so, then it's only a first step towards making Julia output readable on a light background.

File: stdlib/REPL/src/Terminals.jl [hunk 2]
```diff
@@ -169,7 +171,101 @@ else
...
+            end
+        end
+    end
+
+    # Guess from terminal name.
+    # This list of terminal names is taken from vim.
+    if term.term_type == "linux" ||         # Linux console
+        term.term_type == "screen.linux" || # Linux console with screen
+        startswith(term.term_type, "cygwin") || # Cygwin shell
+        startswith(term.term_type, "putty")     # Putty progrem
```
Comment (vtjnash) [inline stdlib/REPL/src/Terminals.jl]: I know this list comes from VIM, but it seems likely ancient. I think most terminals I use report as "xterm" now

File: stdlib/REPL/src/Terminals.jl [hunk 3]
```diff
@@ -169,7 +171,101 @@ else
...
+    if sep !== nothing
+        # See e.g. <https://en.wikipedia.org/wiki/ANSI_escape_code>
+        # for the 16 possible colors
+        colorbg = tryparse(Int, colorfgbg[last(sep)+1:end])
+        if colorbg !== nothing && colorbg >= 0 && colorbg <= 15
+            return colorbg == 7 || colorbg >= 9
+        end
+    end
+
+    # Use OSC 11 escape sequence
```
[the remark this answers is not in the document]
Comment (eschnett) [reply stdlib/REPL/src/Terminals.jl]: Leave out which part – looking at the `COLORFGBG` variable?

File: stdlib/REPL/src/Terminals.jl (hunk 2 above)
Comment (eschnett) [reply stdlib/REPL/src/Terminals.jl]: This is a list of terminals which default to using a dark background. The terminals which default to a light background (`xterm` and friends) are not in this list.

**Comment by vtjnash** [comment]
I think we've somewhat essentially decided not to go with this PR. I think the `--color=light` / `--color=dark` options still have merit however, so I'll keep the corresponding issue open and mention that there

**Comment by StefanKarpinski** [comment]
Probably better as an environment variable since you presumably don't want to have to set it on each invocation.
